aten/src/ATen/native/sparse/cuda/SoftMax.cu

```diff
@@ -0,0 +1,640 @@
+  auto out_values_accessor = out_values_2.packed_accessor<scalar_t, 2>();
+
+  Tensor sorted_indices;
+  Tensor pool_offsets;
+  Tensor pool_sizes;
+  Tensor mx_buffer;
+
+  std::tie(sorted_indices, pool_offsets, pool_sizes, mx_buffer) =
+      compute_pool_max<scalar_t, true>(indices, values_2, sizes, nvalues, dim);
+
+  auto pool_size = pool_offsets.size(0);
+  int block_size = getNumThreads(pool_size);
+  const int grid_size = (pool_size + block_size - 1) / block_size;
+
+  cuda_sparse_coo_softmax_kernel<scalar_t, LogSoftMax>
+      <<<grid_size, block_size, 0, stream>>>(
+          sorted_indices.data_ptr<int64_t>(),
+          pool_size,
+          pool_sizes.data_ptr<int64_t>(),
+          pool_offsets.data_ptr<int64_t>(),
+          nvalues,
+          mx_buffer.data_ptr<scalar_t>(),
+          values_accessor,
+          out_values_accessor);
+  THCudaCheck(cudaGetLastError());
+}
+
+template <typename scalar_t, bool LogSoftMax>
+void cuda_sparse_coo_softmax_backward(
+    Tensor& grad_input,
+    const Tensor& grad,
+    const Tensor& output,
+    const int64_t dim) {
+  /*
+    See ATen/native/sparse/Softmax.cpp:cpu_sparse_coo_softmax_backward for
+    the CPU implementation of the sparse softmax backward algorithm that this
+    implementation is based on.
+  */
+  auto sparse_dim = output.sparse_dim();
+  auto sizes = output.sizes().vec();
+  auto grad_indices = grad._indices().contiguous();
+  auto grad_values = grad._values().contiguous();
+  auto out_indices = output._indices().contiguous();
+  auto out_values = output._values().contiguous();
+  auto values = grad_input._values();
+  auto indices = grad_input._indices();
+  auto out_nnz = out_values.size(0);
+  auto grad_nnz = grad_values.size(0);
+
+  values.resize_as_(out_values);
+  values.zero_();
+  indices.resize_as_(out_indices);
+  indices.copy_(out_indices);
+
+  auto out_offsets = get_offsets(out_indices, sizes, -1);
+  auto grad_offsets = get_offsets(grad_indices, sizes, -1);
+
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  auto allocator = THCThrustAllocator(globalContext().lazyInitCUDA());
+  auto policy = thrust::cuda::par(allocator).on(stream);
+  if (dim >= sparse_dim) {
+    auto host_out_offsets =
+        out_offsets.to(at::Device(kCPU), indices.dtype(), false, true);
+    auto host_grad_offsets =
+        grad_offsets.to(at::Device(kCPU), indices.dtype(), false, true);
+    ;
+    auto out_offsets_accessor = host_out_offsets.data_ptr<int64_t>();
+    auto grad_offsets_accessor = host_grad_offsets.data_ptr<int64_t>();
+    for (int64_t i = 0; i < out_nnz; i++) {
+      Tensor unused = at::native::empty_like(grad_values);
+      auto low = thrust::lower_bound(
+          grad_offsets_accessor,
+          grad_offsets_accessor + grad_offsets.size(0),
+          out_offsets_accessor[i]);
+      auto j = low - grad_offsets_accessor;
+      if (j < grad_nnz && out_offsets_accessor[i] == grad_offsets_accessor[j]) {
```

Comment:
Please add a comment explaining this if check. If this check is false then what happens? Only an empty tensor is returned?